### DISCUSSION TITLE
Syntax for interpret captured variable as constant in Lambda expression as Expression

### DISCUSSION THREAD

**Discussion by RamType0** [dotnet/csharplang#3501] [Language Ideas]
## Backgrounds

Currently, all captured variable in lambda expression are captured by reference.

```C#


int max = GetMax();
var query = db.Data.Where(d=>d.Value<=max);//Closure has been created for max,and Expression references it

```

If we want it as constant,we cannot use lambda expression.

```C#


int max = GetMax();
var param = Expression.Parameter(typeof(DbElement),"d");
var query = db.Data.Where(Expression.Lambda<Func<DbElement,bool>>(Expression.LessThanOrEqual(Expression.MakeMemberAccess(param,typeof(DbElement).GetProperty(nameof(DbElement.Value))),Expression.Constant(max)),param));

```

I want to write this more simply...

## Proposed syntax

```C#


int max = GetMax();
var query = db.Data.Where(d=>d.Value<=(const)max);

```

This is really ugly, and confusing title,but I couldn't figure out any other good title...

**Comment by svick** [comment]
Why does the distinction between a variable captured by reference and a constant make a difference for you?

**Comment by RamType0** [comment]
> Why does the distinction between a variable captured by reference and a constant make a difference for you?

When I compile Expression,switch expression only allows constant for cases.

Also it injects implicit dependency for generated query when working on EntityFrameworkCore.
Generated SQL query string from query with capture by reference never work with alone.

In the first place,Expression is a presentation of expression tree's semantic model.
Different expression tree is different.

**Comment by svick** [comment]
> When I compile Expression,switch expression only allows constant for cases.

You should be able to fairly easily work around that by evaluating the subexpression.

> Also it injects implicit dependency for generated query when working on EntityFrameworkCore.
Generated SQL query string from query with capture by reference never work with alone.

I don't understand what this means, can you clarify? Is the issue that this generates a parametrized query and you don't want that?

> Different expression tree is different.

That is not a good enough reason to add special syntax just for this.

**Comment by Richiban** [comment]
I would hope that readonly locals https://github.com/dotnet/csharplang/issues/188 would allow for this scenario.

**Comment by RamType0** [comment]
> > When I compile Expression,switch expression only allows constant for cases.
> 
> You should be able to fairly easily work around that by evaluating the subexpression.

This problem begins with there is no syntax for injecting sub expression into lambda expression.

**Comment by surfsky** [comment]
Because linq expression is so difficult to write and read, I prefer the  **expressof** expression( see #3499 ), The codes can be beautified  as below:

``` csharp
int max = GetMax();
var exp = expressof(DbElement.Value <= max);
var query = db.Data.Where(ex);
```

**Comment by CyrusNajmabadi** [comment]
> Because linq expression is so difficult to write and read,

I don't see how your version is any less difficult to write or read.

**Comment by theunrepentantgeek** [comment]
> The codes can be beautified as below:

You and I have very different definitions for the word _beautified_.

The C# ecosystem already has good solutions for this problem. That you don’t like them is hardly good motivation for the investment required.

**Comment by RamType0** [comment]
> Because linq expression is so difficult to write and read, I prefer the **expressof** expression( see #3499 ), The codes can be beautified as below:
> 
> ```cs
> int max = GetMax();
> var exp = expressof(DbElement.Value <= max);
> var query = db.Data.Where(ex);
> ```

How is your feature relating this issue?

**Comment by surfsky** [comment]
When I write a extension function with expression like below:

``` csharp
        /// <summary>Get Expression' s Name</summary>
        public static string GetName(this Expression expr)
        {
            if (expr == null)
                return "";
            if (expr is LambdaExpression le)
                return GetName(le.Body);
            if (expr is UnaryExpression ue)
                return GetName((MemberExpression)ue.Operand);
            if (expr is MemberExpression me)
            {
                var name = me.Member.Name;
                if (me.Expression is MemberExpression)
                    return GetName(me.Expression) + "." + name;
                else
                    return name;
            }
            if (expr is ParameterExpression pe)
                return pe.Type.Name;
            return "";
        }
```

Then I write  unit test codes:

``` csharp
        [TestMethod()]
        public void GetNameTest()
        {
            Expression<Func<Person, string>> exp1 = t => t.Name;  // It's so long...
            var exp1 = expressof(Person.Name);          // It's my wish

            Assert.AreEqual(exp1.GetName(), "Name");
        }
```

It's only a simply example. However we will meet other complex expression type, It's hard to write the expression type, so I get this idea.

**Comment by CyrusNajmabadi** [comment]
Why would `var exp1 = expressof(Person.Name); ` be equivalent to `Expression<Func<Person, object>> exp1 = t => t.Name;`?  What are your language rules to make this happen? i.e. can you please specify how `expressof` is actually defined?

**Comment by surfsky** [comment]
``` csharp
nameof(Person.Name)     : Name           // Get name for Person.Name
typeof(Person.Name)     : String         // Get type for Person.Name
expressof(Person.Name)  : t=> t.Name     // Get a expression to fetch Person.Name
```

I don't know how to accomplish it, It's only my syntax proposal .

**Comment by HaloFour** [comment]
`typeof(Person.Name)` isn't legal C# syntax.

`expressof(Person.Name)` seems like it would have very limited use cases.  Doing the same in code today is trivially easy considering how infrequent it is necessary.

**Comment by CyrusNajmabadi** [comment]
> isn't legal C# syntax.

Well, it's legal syntax. It's just semantically not legal :)

**Comment by CyrusNajmabadi** [comment]
> I don't know how to accomplish it, It's only my syntax proposal .

Syntax proposals actually have to state what the syntax *means*.  i'm sill not getting it.  so if i have `expressof(X.Y)` that's just shorthand for `t => t.Y`?

That's it?

**Comment by RamType0** [comment]
@surfsky 
Talk about this issue,please.
You should talk about `expressof` at here. https://github.com/dotnet/csharplang/issues/3499

**Comment by RamType0** [comment]
One of the pros of Expression from delegate is it is white-box.
If you don't need it's semantic structure,delegate is better in most of case.(It is more lightweight,and can be written in multiple expression with lambda expression)

We can optimize it(e.g. `Expression.Reduce`),or implement it on another framework (e.g. SQL query) by analyzing it's semantic structure.

But,captured (by reference) variable in Expression is completely a black-box.
It can be rewritten any times,from anywhere.

Yeah, this analyzing of structure causes issues like serialization.

**Comment by surfsky** [comment]
ok, #3499
